posts/2021-04-21-What I wish somebody told me when I was learning Haskell.md

```diff
@@ -0,0 +1,337 @@
+
+[^productivity]: As the measure of engineering productivity we can use the average time to launch a given set of requirements with a fixed size team, and as the cost of ownership - the engineering time required to maintain the application.
+
+[^fpcomplete]: It is worth watching [the video by the Aaron Contorer](https://www.youtube.com/watch?v=ybSBCVhVWs8), the chairman of FP Complete, about functional programming, where he tells how Haskell allows to achieve productivity, quality and performance without the need to compromise on any of them.
+
+[^diehl]: [What I wish I knew when learning Haskell](http://dev.stephendiehl.com/hask/) is a great writing by Stephen Diehl that gave me a big picture of what can be done with Haskell - for some time it was my go to reference to find other resources.
+
+[^dict]: I did took advantage of Haskell data immutability though, being able to use a dictionary as a key to another dictionary (by value) - in typescript I had to hash it manually.
+
+[^pure]: `pure` and `return` are the same for [well-behaved monads](https://hackage.haskell.org/package/base-4.15.0.0/docs/Control-Monad.html#t:Monad). [Idris](https://www.idris-lang.org) that is very similar to Haskell, with first class dependent types, completely removed `return` and only uses `pure` - you could do it too in your Haskell code.
+
+[^sicp]: A great book on modelling software systems is [Structure and Interpretation of Computer Programs](https://mitpress.mit.edu/sites/default/files/sicp/index.html) that is used as MIT course text - you would become a much better programmer if you study it, regardless of the language you use, and it would help you with the functional programming in particular - it uses Lisp.
+
+[^monad]: "you don't need to understand how it works to play on it" - it is based on the [earlier post](https://byorgey.wordpress.com/2009/01/12/abstraction-intuition-and-the-monad-tutorial-fallacy/) by Brent Yorgey. Some specific monad instances have additional interfaces, but there is a common interface to them all. And you can still play on them differently.
+
+[^lenses]: Lenses are one such example.
+
+[^rich]: I am not rich enough to buy cheap things
+
+[^effects]: There are some advanced composition scenarios that do require using effects, but they are quite rare.
+
+[^n2]: Comparing with big-O notation, the cost to compose effects is linear, but there is a large constant factor, that makes mtl O(n<sup>2</sup>)composition cheaper than linear effects composition in real world applications, particularly given that most of this cost has been covered by mtl library already.
+
+[^msnoyman]: See great posts by Michael Snoyman about [ReaderT pattern](https://www.fpcomplete.com/blog/2017/06/readert-design-pattern/) and [unliftio (a tale of two brackets)](https://www.fpcomplete.com/blog/2017/06/tale-of-two-brackets/)
```

Comment:
The last two footnotes are out of order, should be after `[^monad]` I believe

Reply:
doesn't matter, pandoc replaces them with numbers and puts in the right order.